### PR TITLE
[BEAM-6838] Go SDK: Improving error msgs in pipeline serialization.

### PR DESCRIPTION
**Pull request by youngoli**
This PR is for improving the error messages that appear when forgetting to register a type with Beam before beginning pipeline construction. A specific focus was improving the user experience for when users forget to register Protobuf types.

In addition to that I also updated all the other error messages in the affected files so that they would be roughly consistent and also improved. Mainly this consisted of including information about the input parameters so the user can see what coders/types/etc. are failing to serialize or deserialize.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by youngoli** [comment]
R: @lostluck, @aaltay (committer)

Can you guys give thoughts on if these messages look good (not too wordy, convey useful information, follow Go style conventions), and whether the messages will be formatted in a usable way? I wasn't able to effectively test a lot of the error messages, since I could only see the errors that would trigger with an unregistered type, so I don't know if outputting structs in these error messages a good idea or not.

File: sdks/go/pkg/beam/core/runtime/graphx/coder.go [hunk 1]
```diff
@@ -366,7 +370,7 @@ func (b *CoderMarshaller) Add(c *coder.Coder) string {
 		return b.internBuiltInCoder(urnVarIntCoder)
 
 	default:
-		panic(fmt.Sprintf("Unexpected coder kind: %v", c.Kind))
+		panic(fmt.Sprintf("failed to marshal custom coder %v, unexpected coder kind: %v", *c, c.Kind))
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/graphx/coder.go]: The other panics are getting capitalized (Failed vs failed), please capitalize for consistency.

Errors should always be lowercased due to the concatenation convention. Makes them easier to read.

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go [hunk 2]
```diff
@@ -43,14 +43,14 @@ func EncodeMultiEdge(edge *graph.MultiEdge) (*v1.MultiEdge, error) {
 	if edge.DoFn != nil {
 		ref, err := encodeFn((*graph.Fn)(edge.DoFn))
 		if err != nil {
-			return nil, fmt.Errorf("encode: bad userfn: %v", err)
+			return nil, fmt.Errorf("failed to encode multi edge %v, bad userfn: %v", *edge, err)
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: I'm a bit torn on including "multi edge" since the part that's important is that the user's DoFn isn't able to be encoded, while "multi edge" is an internal detail a pipeline author doesn't run into. My inclination would be to drop "multi edge" and simply say 
("failed to encode userfn %v:", *edge, err)

Including the *edge as context is good since it should help narrow things down for the reading user. Let me know if you want help triggering this message.

Here and the similar messages below.

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go [hunk 3]
```diff
@@ -122,15 +122,15 @@ func DecodeMultiEdge(edge *v1.MultiEdge) (graph.Opcode, *graph.Fn, *window.Fn, [
 func encodeCustomCoder(c *coder.CustomCoder) (*v1.CustomCoder, error) {
 	t, err := encodeType(c.Type)
 	if err != nil {
-		return nil, fmt.Errorf("bad underlying type: %v", err)
+		return nil, fmt.Errorf("failed to encode custom coder %v, bad underlying type: %v", *c, err)
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: I think "underlying type" might end up being misleading, but I can't work out the phraseing.
"failed to encode custom coder %v for type %v: %v", *c,  c.Type,

File: sdks/go/pkg/beam/core/runtime/graphx/coder.go [hunk 4]
```diff
@@ -157,7 +157,7 @@ func (b *CoderUnmarshaller) makeCoder(c *pb.Coder) (*coder.Coder, error) {
 
 	case urnKVCoder:
 		if len(components) != 2 {
-			return nil, fmt.Errorf("bad pair: %v", c)
+			return nil, fmt.Errorf("could not make coder from proto %v, incorrect number of components for pair", *c)
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/graphx/coder.go]: Further, take care when dereferencing pointers for use in printing: Some may have String() methods so you won't have to. See the following as an example: https://play.golang.org/p/WC2gf2wOBNf  
This applies everywhere, even to the protos, which typically have a String() method generated for their pointer types.

[the remark this answers is not in the document]
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/graphx/coder.go]: Useful advice, modified all the error msgs in this function to match.

File: sdks/go/pkg/beam/core/runtime/graphx/coder.go [hunk 5]
```diff
@@ -206,7 +206,7 @@ func (b *CoderUnmarshaller) makeCoder(c *pb.Coder) (*coder.Coder, error) {
 
 	case urnLengthPrefixCoder:
 		if len(components) != 1 {
-			return nil, fmt.Errorf("bad length prefix: %v", c)
+			return nil, fmt.Errorf("could not make coder from proto %v, bad length prefix", *c)
```
[the remark this answers is not in the document]
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/graphx/coder.go]: Done

File: sdks/go/pkg/beam/core/runtime/graphx/coder.go [hunk 6]
```diff
@@ -217,7 +217,7 @@ func (b *CoderUnmarshaller) makeCoder(c *pb.Coder) (*coder.Coder, error) {
 		// the portable pipeline model directly (BEAM-2885)
 		if elm.GetSpec().GetSpec().GetUrn() != "" && elm.GetSpec().GetSpec().GetUrn() != urnCustomCoder {
 			// TODO(herohde) 11/17/2017: revisit this restriction
-			return nil, fmt.Errorf("expected length prefix of custom coder only: %v", elm)
+			return nil, fmt.Errorf("could not make coder from proto %v, expected length prefix of custom coder only: %v", *c, elm)
```
[the remark this answers is not in the document]
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/graphx/coder.go]: Done.

File: sdks/go/pkg/beam/core/runtime/graphx/coder.go [hunk 7]
```diff
@@ -233,7 +233,7 @@ func (b *CoderUnmarshaller) makeCoder(c *pb.Coder) (*coder.Coder, error) {
 
 	case urnWindowedValueCoder:
 		if len(components) != 2 {
-			return nil, fmt.Errorf("bad windowed value: %v", c)
+			return nil, fmt.Errorf("could not make coder from proto %v, bad windowed value", *c)
```
[the remark this answers is not in the document]
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/graphx/coder.go]: Done.

File: sdks/go/pkg/beam/core/runtime/graphx/coder.go [hunk 8]
```diff
@@ -248,7 +248,7 @@ func (b *CoderUnmarshaller) makeCoder(c *pb.Coder) (*coder.Coder, error) {
 		return &coder.Coder{Kind: coder.WindowedValue, T: t, Components: []*coder.Coder{elm}, Window: w}, nil
 
 	case streamType:
-		return nil, fmt.Errorf("stream must be pair value: %v", c)
+		return nil, fmt.Errorf("could not make coder from proto %v, stream must be pair value", *c)
```
[the remark this answers is not in the document]
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/graphx/coder.go]: Done.

File: sdks/go/pkg/beam/core/runtime/graphx/coder.go (hunk 1 above)
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/graphx/coder.go]: Capitalized them all for consistency. I was aware of errors being lowercased, but wasn't sure about panic messages.

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go (hunk 2 above)
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: Done.

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go (hunk 3 above)
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: Yeah, it's a bit tricky. Your suggestion is better than anything I thought of, so I went with it.

File: sdks/go/pkg/beam/core/runtime/graphx/coder.go (hunk 4 above)
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/graphx/coder.go]: That's really useful to know actually. I checked the existing String functions and removed all the dereferences I could (which was all of them).

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go [hunk 9]
```diff
@@ -145,20 +145,20 @@ func encodeCustomCoder(c *coder.CustomCoder) (*v1.CustomCoder, error) {
 func decodeCustomCoder(c *v1.CustomCoder) (*coder.CustomCoder, error) {
 	t, err := decodeType(c.Type)
 	if err != nil {
-		return nil, fmt.Errorf("decodeCustomCoder bad type: %v", err)
+		return nil, fmt.Errorf("failed to decode custom coder proto %v for type %v: %v", c, c.Type, err)
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: Missed removing the "proto " bits throughout this file. I think it's because it's in the second file, understandable oversight. That's why we review.

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go [hunk 10]
```diff
@@ -195,7 +195,7 @@ func encodeFn(u *graph.Fn) (*v1.Fn, error) {
 		gen := reflectx.FunctionName(u.DynFn.Gen)
 		t, err := encodeType(u.DynFn.T)
 		if err != nil {
-			return nil, fmt.Errorf("bad function type: %v", err)
+			return nil, fmt.Errorf("failed to encode function %v, bad function type: %v", u, err)
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: ("failed to encode dynamic DoFn %v, bad function type: %v", u, err)

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go [hunk 11]
```diff
@@ -207,45 +207,45 @@ func encodeFn(u *graph.Fn) (*v1.Fn, error) {
 	case u.Fn != nil:
 		fn, err := encodeUserFn(u.Fn)
 		if err != nil {
-			return nil, fmt.Errorf("bad userfn: %v", err)
+			return nil, fmt.Errorf("failed to encode function %v, bad userfn: %v", u, err)
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: ("failed to encode DoFn %v, bad userfn: %v", u, err)

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go [hunk 12]
```diff
@@ -207,45 +207,45 @@ func encodeFn(u *graph.Fn) (*v1.Fn, error) {
...
 		}
 		return &v1.Fn{Fn: fn}, nil
 
 	case u.Recv != nil:
 		t := reflect.TypeOf(u.Recv)
 		k, ok := runtime.TypeKey(reflectx.SkipPtr(t))
 		if !ok {
-			return nil, fmt.Errorf("bad recv: %v", u.Recv)
+			return nil, fmt.Errorf("failed to encode function %v, bad receiver %v", u, u.Recv)
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: This one is trickier:
("failed to encode structural DoFn %v, bad receiver type: %T", u, u.Recv)

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go [hunk 13]
```diff
@@ -207,45 +207,45 @@ func encodeFn(u *graph.Fn) (*v1.Fn, error) {
...
 		}
 		if _, ok := runtime.LookupType(k); !ok {
-			return nil, fmt.Errorf("recv type must be registered: %v", t)
+			return nil, fmt.Errorf("failed to encode function %v, receiver type %v must be registered", u, t)
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: ("failed to encode structural DoFn %v, receiver type %v must be registered", u, t)

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go [hunk 14]
```diff
@@ -207,45 +207,45 @@ func encodeFn(u *graph.Fn) (*v1.Fn, error) {
...
 		}
 		typ, err := encodeType(t)
 		if err != nil {
-			panic(fmt.Sprintf("bad recv type: %v", u.Recv))
+			panic(fmt.Sprintf("Failed to encode function %v, bad receiver type: %v", u, err))
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: ("Failed to encode structural DoFn %v, bad receiver type %T: %v", u, u.Recv, err)

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go [hunk 15]
```diff
@@ -207,45 +207,45 @@ func encodeFn(u *graph.Fn) (*v1.Fn, error) {
...
 		}
 
 		data, err := json.Marshal(u.Recv)
 		if err != nil {
-			return nil, fmt.Errorf("bad userfn: %v", err)
+			return nil, fmt.Errorf("failed to encode function %v, bad userfn: %v", u, err)
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: ("Failed to encode structural DoFn %v, %T: %v", u, u.Recv, err)

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go [hunk 16]
```diff
@@ -207,45 +207,45 @@ func encodeFn(u *graph.Fn) (*v1.Fn, error) {
...
+		panic(fmt.Sprintf("Failed to encode function %v, empty fn", u))
 	}
 }
 
 func decodeFn(u *v1.Fn) (*graph.Fn, error) {
 	if u.Dynfn != nil {
 		gen, err := runtime.ResolveFunction(u.Dynfn.Gen, genFnType)
 		if err != nil {
-			return nil, fmt.Errorf("bad symbol %v: %v", u.Dynfn.Gen, err)
+			return nil, fmt.Errorf("failed to decode function proto %v, bad symbol %v: %v", u, u.Dynfn.Gen, err)
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: ("failed to decode dynamic DoFn  %v, bad symbol %v: %v", u, u.Dynfn.Gen, err)

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go [hunk 17]
```diff
@@ -207,45 +207,45 @@ func encodeFn(u *graph.Fn) (*v1.Fn, error) {
...
 		}
 
 		t, err := decodeType(u.Dynfn.Type)
 		if err != nil {
-			return nil, fmt.Errorf("bad type: %v", err)
+			return nil, fmt.Errorf("failed to decode function proto %v, bad type: %v", u, err)
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: ("failed to decode dynamic DoFn %v, bad type: %v", u, err)

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go [hunk 18]
```diff
@@ -257,22 +257,22 @@ func decodeFn(u *v1.Fn) (*graph.Fn, error) {
 	if u.Fn != nil {
 		fn, err := decodeUserFn(u.Fn)
 		if err != nil {
-			return nil, fmt.Errorf("bad userfn: %v", err)
+			return nil, fmt.Errorf("failed to decode function proto %v, bad userfn: %v", u, err)
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: ("failed to decode DoFn %v, bad userfn: %v", u, err)

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go [hunk 19]
```diff
@@ -257,22 +257,22 @@ func decodeFn(u *v1.Fn) (*graph.Fn, error) {
...
 		}
 		fx, err := funcx.New(reflectx.MakeFunc(fn))
 		if err != nil {
-			return nil, fmt.Errorf("bad userfn: %v", err)
+			return nil, fmt.Errorf("failed to decode function proto %v, bad userfn: %v", u, err)
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: ("failed to decode DoFn %v, bad userfn: %v", u, err)

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go [hunk 20]
```diff
@@ -257,22 +257,22 @@ func decodeFn(u *v1.Fn) (*graph.Fn, error) {
...
 		}
 		return &graph.Fn{Fn: fx}, nil
 	}
 
 	t, err := decodeType(u.Type)
 	if err != nil {
-		return nil, fmt.Errorf("bad type: %v", err)
+		return nil, fmt.Errorf("failed to decode function proto %v, bad type: %v", u, err)
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: ("failed to decode structural DoFn %v, bad type: %v", u, err)

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go [hunk 21]
```diff
@@ -257,22 +257,22 @@ func decodeFn(u *v1.Fn) (*graph.Fn, error) {
...
 	}
 	fn, err := reflectx.UnmarshalJSON(t, u.Opt)
 	if err != nil {
-		return nil, fmt.Errorf("bad struct encoding: %v", err)
+		return nil, fmt.Errorf("failed to decode function proto %v, bad struct encoding: %v", u, err)
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: ("failed to decode structural DoFn %v, bad userfn: %v", u, err)

Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: Is this a copy/paste error, or did you want "bad struct encoding" changed to "bad userfn"?

Done for the change from "function proto" -> "structural DoFn".

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go (hunk 20 above)
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: Done.

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go (hunk 19 above)
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: Done, with a little added detail:

("failed to decode DoFn %v, failed to construct userfn: %v", u, err)

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go (hunk 18 above)
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: Done with a little added detail:

("failed to decode DoFn %v, failed to decode userfn: %v", u, err)

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go (hunk 17 above)
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: Done.

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go (hunk 16 above)
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: Done.

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go (hunk 15 above)
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: Done, with some changes:

("failed to encode structural DoFn %v, failed to marshal receiver %v: %v", u, u.Recv, err)

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go (hunk 14 above)
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: Done, with a little more detail:

("Failed to encode structural DoFn %v, failed to encode receiver type %T: %v", u, u.Recv, err)

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go (hunk 13 above)
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: Done

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go (hunk 12 above)
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: Done, with a little more detail:

("failed to encode structural DoFn %v, failed to create TypeKey for receiver type %T", u, u.Recv)

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go (hunk 11 above)
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: Done

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go (hunk 10 above)
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: Done.

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go (hunk 9 above)
Comment (youngoli) [reply sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: Done, removed mentions of "proto" in error messages.

**Comment by youngoli** [comment]
There definitely is a lot of repetition with the recursion, even before my changes. Though I'm not sure how to avoid that really. While it's a lot of information for users, it's important for debugging. Only alternative I can think of is adding some debug logs instead of some errors, so users can see the path being taken by examining logs.

Here's a real-world example I've been using to test (I replaced the proto name though)

> failed to encode custom coder: bad underlying type: bad element: bad field type: bad element: bad element: bad field type: bad element: bad field type: bad element: bad element: bad field type: bad field type: bad element: bad field type: unencodable type: protoapi.extensionMap

And the new error:

> Failed to encode custom coder for type protoapi.Message. Make sure the type was registered before calling beam.Init. For example: beam.RegisterType(reflect.TypeOf((*TypeName)(nil)).Elem())
> 
> Full error: failed to encode custom coder *my_package.MyProto[protoapi.Message], bad underlying type: failed to encode pointer *my_package.MyProto, bad base type: failed to encode struct my_package.MyProto, bad field type: failed to encode pointer *my_package.MyProto_MyField, bad base type: failed to encode struct my_package.MyProto_MyField, bad field type: unencodable type map[int32]protoapi.Message

**Comment by lostluck** [comment]
I think the only "root" way to handle this is to have the top level calls (where these are being used non-recursively) summarize the error somehow, which this uses. Having the recursive stack is useful for precision, and it's (as you mention) far better to have more information to explain how/where the error happened. What you're displaying in the last comment is excellent and a clear improvement. We could add newlines if necessary, but I'm also OK just punting to the next version of Gos error improvements. Newlines make some things trickier.

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go (hunk 21 above)
Comment (lostluck) [reply sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: The bad userfn was copypasta, sorry about that. Thanks!
